src/pcp/gadgets.rs

```diff
@@ -49,25 +76,186 @@ impl<F: FieldElement> GadgetCallOnly<F> for Mul<F> {
+            }
+
+            if i < self.poly.len() - 1 {
+                x = poly_mul(&x, inp[0].as_ref());
+            }
+        }
+        Ok(())
+    }
+
+    // Multiply input polynomials using FFT.
+    fn call_poly_fft<V: AsRef<[F]>>(&mut self, outp: &mut [F], inp: &[V]) -> Result<(), PcpError> {
+        let n = self.n;
+        let inp = inp[0].as_ref();
+
+        let mut inp_vals = vec![F::zero(); n];
+        discrete_fourier_transform(&mut inp_vals, inp, n)?;
+
+        let mut x_vals = inp_vals.clone();
+        let mut x = vec![F::zero(); n];
+        x[..inp.len()].clone_from_slice(inp);
+
+        outp[0] = self.poly[0];
+        for i in 1..self.poly.len() {
+            for j in 0..outp.len() {
+                outp[j] += self.poly[i] * x[j];
+            }
+
+            if i < self.poly.len() - 1 {
+                for j in 0..n {
+                    x_vals[j] *= inp_vals[j];
+                }
+
+                discrete_fourier_transform(&mut x, &x_vals, n)?;
+                discrete_fourier_transform_inv_finish(&mut x, n, self.n_inv);
+            }
+        }
+        Ok(())
+    }
+}
+
+impl<F: FieldElement> GadgetCallOnly<F> for PolyEval<F> {
+    fn call(&mut self, inp: &[F]) -> Result<F, PcpError> {
+        gadget_call_check(self, inp.len())?;
+        Ok(poly_eval(&self.poly, inp[0]))
+    }
+
+    fn call_in_len(&self) -> usize {
+        1
+    }
+}
+
+impl<F: FieldElement> Gadget<F> for PolyEval<F> {
+    fn call_poly<V: AsRef<[F]>>(&mut self, outp: &mut [F], inp: &[V]) -> Result<(), PcpError> {
+        gadget_call_poly_check(self, outp, inp)?;
+
+        for i in 0..outp.len() {
+            outp[i] = F::zero();
+        }
+
+        if inp[0].as_ref().len() >= FFT_THRESHOLD_POLY_EVAL {
+            self.call_poly_fft(outp, inp)
+        } else {
+            self.call_poly_direct(outp, inp)
+        }
+    }
+
+    fn call_poly_out_len(&self, in_len: usize) -> usize {
+        poly_deg(&self.poly) * in_len
+    }
+}
+
+/// A wrapper gadget that applies the inner gadget to chunks of input and returns the sum of the
+/// outputs. The arity is equal to the arity of the inner gadget times the number of chunks.
+pub struct ParallelSum<F: FieldElement, G: Gadget<F>> {
+    inner: G,
+    chunks: usize,
+    phantom: PhantomData<F>,
+}
+
+impl<F: FieldElement, G: Gadget<F>> ParallelSum<F, G> {
+    /// Wraps `inner` into a parallel sum gadget with `chunks` chunks.
+    pub fn new(inner: G, chunks: usize, _in_len: usize) -> Self {
+        Self {
+            inner,
+            chunks,
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<F: FieldElement, G: Gadget<F>> GadgetCallOnly<F> for ParallelSum<F, G> {
+    fn call(&mut self, inp: &[F]) -> Result<F, PcpError> {
+        gadget_call_check(self, inp.len())?;
+        let mut outp = F::zero();
+        for chunk in inp.chunks(self.inner.call_in_len()) {
+            outp += self.inner.call(chunk)?;
+        }
```

Comment:
It seems like the computation on each chunk is totally independent. Would it be worthwhile to actually make this parallel? Perhaps we should write up an issue for that future performance optimization.

Reply:
Done: https://github.com/abetterinternet/libprio-rs/issues/34